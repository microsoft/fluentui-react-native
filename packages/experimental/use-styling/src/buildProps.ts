import { GetMemoValue } from '@fluentui-react-native/memo-cache';

/**
 * Denotes what props will be treated as tokens
 * - false | undefined : this means no properties should be treated as tokens
 * - true              : treat all props as tokens
 * - array of keys     : props matching that array should be treated as tokens
 */
export type TokenPropMask<TTokens> = (keyof TTokens)[] | boolean;

/**
 * Raw format for producing styles in a functional manner. These can only depend on tokens or theme as inputs.
 * - tokens:  these will be produced from the theme and component constants, then they will be potentially
 *            be modified by the props. See the TokenPropMask type for more details.
 * - theme:   the theme is provided for reference
 * The provided
 * cache will be scoped to the theme, slot, and tokens that are coming out of the theme.
 */
export type BuildPropsBase<TProps, TTokens, TTheme> = (tokens: TTokens, theme: TTheme, cache: GetMemoValue<any>) => Partial<TProps>;

/**
 * A refine function allows style functions to be updated based on tokens that are also props. Only those tokens that are also
 * props need to be considered as a key for caching
 */
export type RefineFunctionBase<TProps, TTokens, TTheme> = (mask?: TokenPropMask<TTokens>) => BuildPropsBase<TProps, TTokens, TTheme>;

/**
 * Signature for a style function which can be optionally refined by the styling hook if prop masks are provided
 */
export type RefinableBuildPropsBase<TProps, TTokens, TTheme> = BuildPropsBase<TProps, TTokens, TTheme> & {
  refine?: RefineFunctionBase<TProps, TTokens, TTheme>;
};

/**
 * Style functions can be plain functions, refinable functions, or just raw props
 */
export type BuildSlotProps<TSlotProps, TTokens, TTheme> = {
  [K in keyof TSlotProps]?: RefinableBuildPropsBase<TSlotProps[K], TTokens, TTheme> | TSlotProps[K]
};

function cacheStyleClosure<TProps, TTokens, TTheme>(
  fn: (tokens: TTokens, theme: TTheme) => TProps,
  keys?: (keyof TTokens)[]
): RefinableBuildPropsBase<TProps, TTokens, TTheme> {
  return (tokens: TTokens, theme: TTheme, cache: GetMemoValue<TProps>) =>
    cache(() => fn(tokens, theme), (keys || []).map(key => tokens[key]))[0];
}

function refineKeys<TTokens>(keys: (keyof TTokens)[], mask?: TokenPropMask<TTokens>): (keyof TTokens)[] {
  return typeof mask === 'object' && Array.isArray(mask) ? keys.filter(key => mask.findIndex(val => val === key) !== -1) : mask ? keys : [];
}

/**
 * Standard wrapper for a function that provides props for a component based on tokens and theme.
 *
 * @param fn - function which does the work of producing props for the tokens and theme provided
 * @param keys - which token properties are used by this style, this determines the keys to use for caching
 */
export function buildProps<TProps, TTokens, TTheme>(
  fn: (tokens: TTokens, theme: TTheme) => TProps,
  keys?: (keyof TTokens)[]
): RefinableBuildPropsBase<TProps, TTokens, TTheme> {
  // wrap the provided function in the standard caching layer, basing it upon the provided keys
  const result = cacheStyleClosure(fn, keys);

  // if results are being cached on keys, provide the ability to refine the function if a prop mask is specified
  result.refine =
    keys && keys.length > 0
      ? (mask?: TokenPropMask<TTokens>) => {
          return cacheStyleClosure(fn, refineKeys(keys, mask));
        }
      : undefined;

  // return the style function decorated with the refine function
  return result;
}

/**
 * Utility function to check the type and refinement capabilities of a styleFunction and refine it if appropriate
 *
 * @param fn - function or props to potentially refine
 * @param mask - prop mask to use for refinement
 */
export function refinePropsFunctions<TSlotProps, TTokens, TTheme>(
  styles: BuildSlotProps<TSlotProps, TTokens, TTheme>,
  mask: TokenPropMask<TTokens>
): BuildSlotProps<TSlotProps, TTokens, TTheme> {
  const result = {};
  Object.keys(styles).forEach(key => {
    const refine =
      typeof styles[key] === 'function' && (styles[key] as RefinableBuildPropsBase<TSlotProps[keyof TSlotProps], TTokens, TTheme>).refine;
    result[key] = refine ? refine(mask) : styles[key];
  });
  return result;
}
