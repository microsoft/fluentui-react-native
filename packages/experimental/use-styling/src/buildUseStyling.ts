import { getMemoCache, GetMemoValue } from '@fluentui-react-native/memo-cache';
import { immutableMerge } from '@fluentui-react-native/immutable-merge';
import { TokensThatAreAlsoProps, BuildSlotProps, refinePropsFunctions } from './buildProps';
import { HasLayer, applyTokenLayers } from './applyTokenLayers';

/** A function to generate tokens based on a theme */
export type TokensFromTheme<TTokens, TTheme> = (theme: TTheme) => TTokens;

/**
 * Types of tokens, can be:
 * - string - will lookup the name in the theme
 * - Tokens - will merge the tokens in directly
 * - Function - will run against the theme once for each unique theme encountered
 */
export type TokenSettings<TTokens, TTheme> = string | TTokens | TokensFromTheme<TTokens, TTheme>;

/**
 * Options used to build up a useStyling hook
 */
export type UseStylingOptions<TProps, TSlotProps, TTokens, TTheme> = {
  /**
   * Baseline tokens for this component
   */
  tokens?: TokenSettings<TTokens, TTheme>[];

  /**
   * States that might be applied for the component like disabled or hovered, these should be listed
   * in the order that they should be applied
   */
  states?: (keyof TTokens)[];

  /**
   * Functions which build up the props for each slot
   */
  slotProps?: BuildSlotProps<TSlotProps, TTokens, TTheme>;

  /**
   * Which props should be considered to be tokens.
   * - If an array of keys this will ensure these props are promoted to tokens
   * - If true all props will be added to tokens, if false or not specified no props will be treated as tokens
   */
  tokensThatAreAlsoProps?: TokensThatAreAlsoProps<TTokens>;

  /** purely used to make type inferencing work correctly so the hook builder can pick up TProps from this type */
  _propsType?: TProps;
};

/**
 * Signature for the use styling hook
 */
export type UseStyling<TProps, TSlotProps> = (props: TProps, lookup?: HasLayer) => TSlotProps;

/**
 * Helper object which injects theme specific functionality
 */
export type ThemeHelper<TTheme> = {
  /** query the theme from the context, or from a global if your system doesn't use theming */
  useTheme: () => TTheme;

  /** lookup info for the component in the theme */
  getComponentInfo: (theme: TTheme, name: string) => any;
};

/**
 * This the merges [all/some/no] props into tokens based on tokenProps being [true/[key1, key2]/false] respectively. If
 * tokens is modified it returns a new object rather than modifying the tokens object itself
 *
 * @param tokens - set of tokens that are coming from the theme
 * @param props - input props for the component
 * @param tokenProps - how to merge props into tokens. If true all props will be merged, false: no props, an array of keys will
 *                     be treated as an include mask
 */
function promotePropsToTokens<TTokens, TProps>(tokens: TTokens, props: TProps, tokenProps?: TokensThatAreAlsoProps<TTokens>): TTokens {
  return tokenProps && tokenProps !== 'none'
    ? {
        ...tokens,
        ...(typeof tokenProps === 'object' && Array.isArray(tokenProps)
          ? Object.assign(
              {},
              ...tokenProps.filter((key) => props[key as string] !== undefined).map((key) => ({ [key]: props[key as string] })),
            )
          : props),
      }
    : tokens;
}

/**
 * Tokens are defined as either:
 *   TTokens     - an object
 *   string      - a name to look up in the theme
 *   function    - a function to run against the theme to produce tokens
 *
 * This function maps any of these types into a specific TTokens object.  A string is first lookup up in the theme, returning a function
 * or object. If the type is a function this will be invoked with the theme to generate the tokens object.
 *
 * @param tokenEntry - token entry to start with
 * @param theme - theme to use for queries
 * @param getComponentInfo - helper to use to lookup the component in the theme
 */
function mapToTokens<TTokens, TTheme>(
  tokenEntry: TTokens | string | TokensFromTheme<TTokens, TTheme>,
  theme: TTheme,
  getComponentInfo: ThemeHelper<TTheme>['getComponentInfo'],
): object {
  if (typeof tokenEntry === 'string') {
    tokenEntry = getComponentInfo(theme, tokenEntry);
  }
  if (typeof tokenEntry === 'function') {
    tokenEntry = (tokenEntry as TokensFromTheme<TTokens, TTheme>)(theme);
  }
  return (tokenEntry as unknown) as object;
}

/**
 * Produce the final slot props for the styled hook
 *
 * @param styles - refined style functions or props to use for processing
 * @param tokens - token inputs for the style functions
 * @param theme - theme to resolve against
 * @param cache - cache to use for the base of slot caching
 */
function resolveToSlotProps<TSlotProps, TTokens, TTheme>(
  styles: BuildSlotProps<TSlotProps, TTokens, TTheme>,
  tokens: TTokens,
  theme: TTheme,
  cache: GetMemoValue<TTokens>,
): TSlotProps {
  const slotProps = {};
  Object.keys(styles).forEach((key) => {
    const style = styles[key];
    slotProps[key] = typeof style === 'function' ? style(tokens, theme, cache(null, [key])[1]) : style;
  });
  return slotProps as TSlotProps;
}

/**
 * Construct a useStyling hook which returns styled slot props based on props and tokens defined in options and in the theme
 *
 * @param options - options which drive behavior for the generated styling hook
 * @param themeHelper - injected theme functionality
 */
export function buildUseStyling<TProps, TSlotProps, TTokens, TTheme>(
  options: UseStylingOptions<TProps, TSlotProps, TTokens, TTheme>,
  themeHelper: ThemeHelper<TTheme>,
): UseStyling<TProps, TSlotProps> {
  // create a cache instance for this use styling implementation
  const cache = getMemoCache();
  const { useTheme, getComponentInfo } = themeHelper;
  const { tokens, tokensThatAreAlsoProps: tokenProps } = options;
  const styles = refinePropsFunctions(options.slotProps || {}, tokenProps);

  return (props: TProps, lookup?: HasLayer) => {
    // query the theme
    const theme = useTheme();

    // get the base styles all merged together, these will only depend on internal tokens and theme
    const [merged, subCache] = cache(() => immutableMerge(...tokens.map((value) => mapToTokens(value, theme, getComponentInfo))), [theme]);

    // resolve overrides as appropriate
    const [layered, cacheBase] = options.states
      ? applyTokenLayers(merged, options.states as string[], subCache, lookup || ((val) => props[val]))
      : [merged, subCache];

    // now resolve tokens
    const finalTokens = promotePropsToTokens(layered, props, tokenProps);

    // finally produce slotProps from calling the style functions on each entry
    return resolveToSlotProps(styles, finalTokens, theme, cacheBase);
  };
}
