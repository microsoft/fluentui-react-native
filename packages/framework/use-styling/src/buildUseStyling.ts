import { GetMemoValue } from '@fluentui-react-native/memo-cache';
import { TokensThatAreAlsoProps, BuildSlotProps, refinePropsFunctions } from './buildProps';
import { applyPropsToTokens, applyTokenLayers, buildUseTokens, HasLayer, TokenSettings } from '@fluentui-react-native/use-tokens';

/**
 * Options used to build up a useStyling hook
 */
export type UseStylingOptions<TProps, TSlotProps, TTokens, TTheme> = {
  /**
   * Baseline tokens for this component
   */
  tokens?: TokenSettings<TTokens, TTheme>[];

  /**
   * States that might be applied for the component like disabled or hovered, these should be listed
   * in the order that they should be applied
   */
  states?: (keyof TTokens)[];

  /**
   * Functions which build up the props for each slot
   */
  slotProps?: BuildSlotProps<TSlotProps, TTokens, TTheme, TProps>;

  /**
   * Which props should be considered to be tokens.
   * - If an array of keys this will ensure these props are promoted to tokens
   * - If true all props will be added to tokens, if false or not specified no props will be treated as tokens
   */
  tokensThatAreAlsoProps?: TokensThatAreAlsoProps<TTokens>;

  /** purely used to make type inferencing work correctly so the hook builder can pick up TProps from this type */
  _propsType?: TProps;
};

/**
 * Signature for the use styling hook
 */
export type UseStyling<TProps, TSlotProps> = (props: TProps, lookup?: HasLayer) => TSlotProps;

/**
 * Helper object which injects theme specific functionality
 */
export type ThemeHelper<TTheme> = {
  /** query the theme from the context, or from a global if your system doesn't use theming */
  useTheme: () => TTheme;

  /** lookup info for the component in the theme */
  getComponentInfo: (theme: TTheme, name: string) => any;
};

/**
 * Produce the final slot props for the styled hook
 *
 * @param styles - refined style functions or props to use for processing
 * @param tokens - token inputs for the style functions
 * @param theme - theme to resolve against
 * @param props - props from outer component
 * @param cache - cache to use for the base of slot caching
 */
function resolveToSlotProps<TSlotProps, TTokens, TTheme, TProps>(
  styles: BuildSlotProps<TSlotProps, TTokens, TTheme, TProps>,
  tokens: TTokens,
  theme: TTheme,
  props: TProps,
  cache: GetMemoValue<TTokens>,
): TSlotProps {
  const slotProps = {};
  Object.keys(styles).forEach(key => {
    const style = styles[key];
    slotProps[key] = typeof style === 'function' ? style(tokens, theme, props, cache(null, [key])[1]) : style;
  });
  return slotProps as TSlotProps;
}

/**
 * Construct a useStyling hook which returns styled slot props based on props and tokens defined in options and in the theme
 *
 * @param options - options which drive behavior for the generated styling hook
 * @param themeHelper - injected theme functionality
 */
export function buildUseStyling<TProps, TSlotProps, TTokens, TTheme>(
  options: UseStylingOptions<TProps, TSlotProps, TTokens, TTheme>,
  themeHelper: ThemeHelper<TTheme>,
): UseStyling<TProps, TSlotProps> {
  // create a cache instance for this use styling implementation
  const { useTheme, getComponentInfo } = themeHelper;
  const { tokens, tokensThatAreAlsoProps: tokenProps } = options;
  const styles = refinePropsFunctions(options.slotProps || {}, tokenProps);
  const useTokens = buildUseTokens<TTokens, TTheme>(getComponentInfo, ...tokens);

  return (props: TProps, lookup?: HasLayer) => {
    // query the theme
    const theme = useTheme();

    // get the merged tokens from the theme
    let [mergedTokens, cache] = useTokens(theme);

    // resolve overrides as appropriate
    if (options.states) {
      [mergedTokens, cache] = applyTokenLayers(mergedTokens, options.states as string[], cache, lookup || (val => props[val]));
    }

    // now resolve tokens
    if (typeof tokenProps === 'object' && Array.isArray(tokenProps)) {
      [mergedTokens, cache] = applyPropsToTokens(props, mergedTokens, cache, tokenProps);
    } else if (tokenProps === 'all') {
      mergedTokens = { ...mergedTokens, ...props };
    }

    // finally produce slotProps from calling the style functions on each entry
    return resolveToSlotProps(styles, mergedTokens, theme, props, cache);
  };
}
