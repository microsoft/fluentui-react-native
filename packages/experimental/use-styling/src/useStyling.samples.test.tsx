import { ThemeHelper, buildUseStyling, UseStylingOptions } from './buildUseStyling';
import { TextProps, Text, View, ColorValue } from 'react-native';
import { buildProps } from './buildProps';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { mount } from 'enzyme';

/**
 * Sample super simple theming implementation, shared by all the samples. This is intended to be illustrative,
 * not to be production ready theming code
 */

/**
 * The theme just contains global values and options for looking up component overrides
 */
type Theme = {
  globals: {
    backgroundColor: ColorValue;
    color: ColorValue;
    borderColor: ColorValue;
    fontFamily: string;
    fontSize: number;
  };
  components: {
    [key: string]: any;
  };
};

/**
 * The default/base theme just contains base values
 */
const baseTheme: Theme = {
  globals: {
    backgroundColor: 'white',
    color: 'black',
    borderColor: 'blue',
    fontFamily: 'Arial',
    fontSize: 12,
  },
  components: {},
};

const current = { theme: baseTheme };

/**
 * this wrapper solves the (so-far) inexplicable type errors from the matchers in typescript
 */
function snapshotTestTree(tree: any) {
  (expect(toJson(tree)) as any).toMatchSnapshot();
}

/**
 * Because the buildUseStyling utility is not opinionated about theming, the theming is injected via
 * a theme helper implementation. This allows for a framework to build an opinionated version with
 * whatever system they desire
 */
const themeHelper: ThemeHelper<Theme> & { setActive: (theme?: Partial<Theme>) => void } = {
  useTheme: () => current.theme,
  getComponentInfo: (theme: Theme, name: string) => theme.components[name],
  setActive: (theme?: Partial<Theme>) => {
    current.theme = theme ? { ...baseTheme, ...theme } : baseTheme;
  },
};

describe('useStyling samples', () => {
  /**
   * Sample #1 - Themeable text element
   *
   * This adds some default opinions for how a text element should be styled but only allows for customization
   * via theming
   */

  // the component itself is just text props with no style allowed
  type Sample1Props = Omit<TextProps, 'style'>;

  // internally the component just wraps a Text element
  type Sample1SlotProps = { content: TextProps };

  // the tokens for customization are just the color and font props from the theme
  type Sample1Tokens = {
    color?: ColorValue;
    fontFamily?: string;
    fontSize?: number;
  };

  // now create the styling hook, first the options so they can be reused later
  const sample1StylingOptions: UseStylingOptions<Sample1Props, Sample1SlotProps, Sample1Tokens, Theme> = {
    /**
     * tell the styling hook how to build up the tokens
     */
    tokens: [
      /** first the default values should come from the global theme section */
      (t: Theme) => ({
        color: t.globals.color,
        fontFamily: t.globals.fontFamily,
        fontSize: t.globals.fontSize,
      }),
      /** next we should look for a component reference to overlay */
      'Sample1',
    ],
    /**
     * Now provide the recipe for how to build props for the sub-components given the tokens
     */
    slotProps: {
      /** only one sub-component, a Text element called content, as a result this needs to build up TextProps */
      content: buildProps(
        /**
         * first input for buildProps is a function which takes tokens and a theme and returns props
         */
        (tokens: Sample1Tokens /*, theme: Theme */) => {
          return {
            style: { ...tokens },
          };
        },
        /**
         * The second input are the tokens used as inputs for the above function. This is similar to the way the useEffect hook
         * works in react.
         */
        ['color', 'fontFamily', 'fontSize'],
      ),
    },
  };

  // now build the actual hook from the options and theme helper
  const useStylingSample1 = buildUseStyling(sample1StylingOptions, themeHelper);

  // now the sample 1 component becomes simple to build, just merge the styled props with the input props
  const Sample1Text: React.FunctionComponent<Sample1Props> = (props) => {
    const styledProps = useStylingSample1(props).content;
    const merged = { ...props, ...styledProps };
    return <Text {...merged}>{props.children}</Text>;
  };

  /** first render the component with no updates */
  it('Sample1Text rendering with no overrides', () => {
    const tree = mount(<Sample1Text>Sample1a</Sample1Text>);
    snapshotTestTree(tree);
  });

  /** now re-theme the component via the components in the theme */
  it('Sample1Text rendering with some custom settings in the theme', () => {
    themeHelper.setActive({
      components: {
        Sample1: {
          color: 'pink',
          fontSize: 24,
        },
      },
    });
    const tree = mount(<Sample1Text>Sample1b</Sample1Text>);
    snapshotTestTree(tree);
  });

  /**
   * Sample 2 - take the styled text component from the first example and allow color to be set in props as well. This is
   * effectively saying that all other styles come from the theme except that color can be overriden via props.
   *
   * This will start with the definition from sample 1 and then extend it.
   */

  /** use the props from the first sample and just add a color setting */
  type Sample2Props = Sample1Props & { color?: string };

  /**
   * Build the styling hook for sample2. Because this isn't being recombined this is being specified inline rather
   * than using a separate options object. Both are fine.
   */
  const useStylingSample2 = buildUseStyling<Sample2Props, Sample1SlotProps, Sample1Tokens, Theme>(
    {
      /**
       * This just starts with the baseline styling from sample1, in particular we are using the recipes of how to turn
       * token values into the props for the internal sub-components. While this example is not super complex, for real-world
       * components, re-using these can be extremely valuable.
       *
       * With that in mind, this copies over the recipes for how to turn tokens into props, the customizations that
       * will be made are about how to ensure the tokens are set up correctly.
       */
      ...sample1StylingOptions,
      /**
       * In sample1 tokens are set to defaults from the global theme section, then patched with any values looked up with
       * the string 'Sample1'
       *
       * We want to maintain the logic of setting up the globals, but add an additional lookup for 'Sample2'. This might correspond
       * to saying that if we were making a variant of a 'Text' component called 'HeaderText', we might want to look up
       * customizations from 'Text' first, then override those customizations with those from 'HeaderText'
       *
       * If we didn't want to add the extra 'Sample2' lookup this line would be omitted. If we didn't want to look up 'Sample1' first
       * that could be filtered out of the array that is being copied
       */
      tokens: [...sample1StylingOptions.tokens, 'Sample2'],
      /**
       * This is the final bit of magic. The tokens will already have values set from the global theme, they will then be patched with
       * any customizations set into Sample1 and/or Sample2.
       *
       * If this value was omitted then the tokens would be passed to the slotProps recipies as is. To have those values patched from
       * the component props we add a list of the props which need to be passed into tokens. If all props should be spread into the
       * tokens then this value can be set to 'all'. If none should be passed it can be omitted or set to 'none'
       */
      tokensThatAreAlsoProps: ['color'],
    },
    themeHelper,
  );

  // the Sample2Text component is built the same way as sample1, just using the new hook that has been created
  const Sample2Text: React.FunctionComponent<Sample2Props> = (props) => {
    const styledProps = useStylingSample2(props).content;
    const merged = { ...props, ...styledProps };
    // delete the color key to not pass it through to the text props, could be done via destructuring, filtering, or any number of ways
    delete merged.color;
    // render the text
    return <Text {...merged}>{props.children}</Text>;
  };

  /** rendering the Sample2 component with the base theme */
  it('Sample2Text rendering with defaults and a color override', () => {
    themeHelper.setActive();
    const tree = mount(
      <View>
        <Sample2Text>Sample2 with defaults</Sample2Text>
        <Sample2Text color="green">Sample2 with color override via prop</Sample2Text>
      </View>,
    );
    snapshotTestTree(tree);
  });

  /** now re-theme the component via the components in the theme */
  it('Sample2Text rendering with some custom settings in the theme', () => {
    themeHelper.setActive({
      components: {
        Sample1: {
          color: 'pink',
          fontSize: 24,
        },
        Sample2: {
          fontSize: 18,
          fontFamily: 'Helvetica',
        },
      },
    });
    const tree = mount(
      <View>
        <Sample2Text>Sample2 with theme overrides set</Sample2Text>
        <Sample2Text color="purple">Sample2 with theme and color prop override</Sample2Text>
      </View>,
    );
    snapshotTestTree(tree);
  });
});
