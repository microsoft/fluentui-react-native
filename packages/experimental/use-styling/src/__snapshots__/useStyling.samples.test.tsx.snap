// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useStyling samples Sample1Text rendering with no overrides 1`] = `
<Sample1Text>
  <Text
    style={
      Object {
        "color": "black",
        "fontFamily": "Arial",
        "fontSize": 12,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "black",
          "fontFamily": "Arial",
          "fontSize": 12,
        }
      }
    >
      Sample1a
    </Text>
  </Text>
</Sample1Text>
`;

exports[`useStyling samples Sample1Text rendering with some custom settings in the theme 1`] = `
<Sample1Text>
  <Text
    style={
      Object {
        "color": "pink",
        "fontFamily": "Arial",
        "fontSize": 24,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "pink",
          "fontFamily": "Arial",
          "fontSize": 24,
        }
      }
    >
      Sample1b
    </Text>
  </Text>
</Sample1Text>
`;

exports[`useStyling samples Sample2Text rendering with defaults and a color override 1`] = `
<View>
  <View>
    <Sample2Text>
      <Text
        style={
          Object {
            "color": "black",
            "fontFamily": "Arial",
            "fontSize": 12,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "black",
              "fontFamily": "Arial",
              "fontSize": 12,
            }
          }
        >
          Sample2 with defaults
        </Text>
      </Text>
    </Sample2Text>
    <Sample2Text
      color="green"
    >
      <Text
        style={
          Object {
            "color": "green",
            "fontFamily": "Arial",
            "fontSize": 12,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "green",
              "fontFamily": "Arial",
              "fontSize": 12,
            }
          }
        >
          Sample2 with color override via prop
        </Text>
      </Text>
    </Sample2Text>
  </View>
</View>
`;

exports[`useStyling samples Sample2Text rendering with some custom settings in the theme 1`] = `
<View>
  <View>
    <Sample2Text>
      <Text
        style={
          Object {
            "color": "pink",
            "fontFamily": "Helvetica",
            "fontSize": 18,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "pink",
              "fontFamily": "Helvetica",
              "fontSize": 18,
            }
          }
        >
          Sample2 with theme overrides set
        </Text>
      </Text>
    </Sample2Text>
    <Sample2Text
      color="purple"
    >
      <Text
        style={
          Object {
            "color": "purple",
            "fontFamily": "Helvetica",
            "fontSize": 18,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "purple",
              "fontFamily": "Helvetica",
              "fontSize": 18,
            }
          }
        >
          Sample2 with theme and color prop override
        </Text>
      </Text>
    </Sample2Text>
  </View>
</View>
`;
